File: Fundamentals_of_Software_Design/2_Algorithms/Поиск подстроки в строке алгоритмом Боуэра-Мура/main.py
def preprocess_bad_character_rule(needle):
    """Предобработка для правила плохих символов"""
    bad_char_table = {}
    m = len(needle)

    for i in range(m):
        bad_char_table[needle[i]] = i  # Индекс последнего появления символа в needle

    return bad_char_table

def my_strstr(haystack, needle):
    n = len(haystack)
    m = len(needle)

    if m == 0:
        return 0  # Пустая строка всегда найдется

    bad_char_table = preprocess_bad_character_rule(needle)

    i = 0
    while i <= n - m:
        j = m - 1
        while j >= 0 and haystack[i + j] == needle[j]:
            j -= 1

        if j < 0:
            return i  # Возвращаем индекс начала совпадения

        # Сдвигаем на основе таблицы плохих символов
        shift = j - bad_char_table.get(haystack[i + j], -1)
        i += max(shift, 1)

    return -1  # Если подстрока не найдена

# Тестирование
haystack = "this is a simple example"
needle = "simple"

result = my_strstr(haystack, needle)

if result != -1:
    print(f"Подстрока найдена на позиции: {result}")
else:
    print("Подстрока не найдена")

File: Fundamentals_of_Software_Design/2_Algorithms/Поиск подстроки в строке алгоритмом Боуэра-Мура/main.c
#include <stdio.h>
#include <string.h>

#define ALPHABET_SIZE 256  // Размер алфавита (ASCII)

void preprocess_bad_character_rule(const char *needle, int *bad_char_table) {
    int m = strlen(needle);

    // Инициализируем таблицу плохих символов (-1 означает, что символ не найден в паттерне)
    for (int i = 0; i < ALPHABET_SIZE; i++) {
        bad_char_table[i] = -1;
    }

    // Заполняем таблицу плохих символов
    for (int i = 0; i < m; i++) {
        bad_char_table[(unsigned char)needle[i]] = i;
    }
}

int my_strstr(const char *haystack, const char *needle) {
    int n = strlen(haystack);
    int m = strlen(needle);

    if (m == 0) {
        return 0;  // Пустая строка всегда найдется в любой строке
    }

    int bad_char_table[ALPHABET_SIZE];
    preprocess_bad_character_rule(needle, bad_char_table);

    int i = 0;
    while (i <= n - m) {
        int j = m - 1;

        // Сравниваем подстроку haystack с needle, начиная с конца needle
        while (j >= 0 && haystack[i + j] == needle[j]) {
            j--;
        }

        // Если нашли полное совпадение
        if (j < 0) {
            return i;  // Возвращаем индекс начала совпадения
        }

        // Сдвигаем паттерн в соответствии с таблицей плохих символов
        i += (j - bad_char_table[(unsigned char)haystack[i + j]] > 0) ?
             j - bad_char_table[(unsigned char)haystack[i + j]] : 1;
    }

    return -1;  // Если подстрока не найдена
}

int main() {
    const char *haystack = "this is a simple example";
    const char *needle = "simple";

    int result = my_strstr(haystack, needle);

    if (result != -1) {
        printf("Подстрока найдена на позиции: %d\n", result);
    } else {
        printf("Подстрока не найдена\n");
    }

    return 0;
}
